File: TwoWayDatabinding/app/src/main/java/com/enpassio/twowaydatabinding/ui/ToyAdapter.java
package com.enpassio.twowaydatabinding.ui;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.enpassio.twowaydatabinding.R;
import com.enpassio.twowaydatabinding.data.model.ToyEntry;
import com.enpassio.twowaydatabinding.databinding.ItemToyBinding;

import java.util.List;

public class ToyAdapter extends RecyclerView.Adapter<ToyAdapter.ToyViewHolder>{

    private List<ToyEntry> mToyList;
    private final ToyClickListener mListener;

    ToyAdapter(ToyClickListener listener) {
        mListener = listener;
    }

    void setToyList(List<ToyEntry> articleList) {
        mToyList = articleList;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ToyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemToyBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.item_toy,
                        parent, false);
        return new ToyViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ToyViewHolder holder, int position) {
       holder.bind(mToyList.get(position), mListener);
    }

    @Override
    public int getItemCount() {
        //If list is null, return 0, otherwise return the size of the list
        return mToyList == null ? 0 : mToyList.size();
    }

    class ToyViewHolder extends RecyclerView.ViewHolder{

        final ItemToyBinding binding;

        ToyViewHolder(@NonNull ItemToyBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(ToyEntry currentToy, ToyClickListener clickListener){
            //For each item, corresponding product object is passed to the binding
            binding.setToy(currentToy);
            //Pass an item click listener to each item layout.
            binding.setToyItemClick(mListener);
            //This is to force bindings to execute right away
            binding.executePendingBindings();
        }
    }

    public interface ToyClickListener {
        void onToyClicked(ToyEntry chosenToy);
    }
}

File: TwoWayDatabindingKotlin/app/src/main/java/com/enpassion/twowaydatabindingkotlin/ui/ToyAdapter.kt
package com.enpassion.twowaydatabindingkotlin.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.enpassion.twowaydatabindingkotlin.R
import com.enpassion.twowaydatabindingkotlin.data.ToyEntry
import com.enpassion.twowaydatabindingkotlin.databinding.ItemToyBinding


class ToyAdapter(private val mListener: ToyClickListener) :
    RecyclerView.Adapter<ToyAdapter.ToyViewHolder>() {

    var toyList: List<ToyEntry>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ToyViewHolder = ToyViewHolder.from(parent)

    override fun onBindViewHolder(holder: ToyViewHolder, position: Int) = holder.bind(toyList?.get(position), mListener)

    override fun getItemCount(): Int {
        //If list is null, return 0, otherwise return the size of the list
        return toyList?.size ?: 0
    }

    class ToyViewHolder(private val binding: ItemToyBinding) : RecyclerView.ViewHolder(binding.root){

        fun bind(currentToy: ToyEntry?, clickListener : ToyClickListener){
            binding.toy = currentToy
            binding.toyItemClick = clickListener
            binding.executePendingBindings()
        }

        companion object {
            fun from(parent: ViewGroup): ToyViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = DataBindingUtil
                    .inflate<ItemToyBinding>(layoutInflater, R.layout.item_toy,
                        parent, false)
                return ToyViewHolder(binding)
            }
        }
    }

    interface ToyClickListener {
        fun onToyClicked(chosenToy: ToyEntry)
    }

}